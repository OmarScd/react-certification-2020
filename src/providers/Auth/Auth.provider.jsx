import React, { useState, useEffect, useContext } from 'react';

import { USERNAME_STORAGE_KEY } from '../../utils/constants';
import { storage } from '../../utils/storage';

const AuthContext = React.createContext(null);

const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error(`Can't use "useAuth" without an AuthProvider!`);
  }
  return context;
};

const AuthProvider = ({ children }) => {
  const [authenticated, setAuthenticated] = useState(false);
  const [username, setUsername] = useState(undefined);

  useEffect(() => {
    const user = storage.get(USERNAME_STORAGE_KEY);
    const isAuthenticated = !!user;

    setAuthenticated(isAuthenticated);
    if (isAuthenticated) {
      setUsername(user);
    }
  }, []);

  const login = (name) => {
    setAuthenticated(true);
    setUsername(name);
    storage.set(USERNAME_STORAGE_KEY, name);
  };

  const logout = () => {
    setAuthenticated(false);
    setUsername(undefined);
    storage.remove(USERNAME_STORAGE_KEY);
  };

  return (
    <AuthContext.Provider value={{ login, logout, authenticated, username }}>
      {children}
    </AuthContext.Provider>
  );
};

export { useAuth, AuthProvider as default };
