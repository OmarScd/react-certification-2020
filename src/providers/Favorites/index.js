export { default } from './Favorites.provider';
