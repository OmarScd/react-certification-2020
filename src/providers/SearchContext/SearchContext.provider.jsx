import React, { useState, useContext } from 'react';

const SearchContext = React.createContext();

const useSearch = () => {
  const context = useContext(SearchContext);
  if (!context) {
    throw new Error(`Can't use "useSearch" without a FavoritesProvider!`);
  }
  return context;
};

const SearchContextProvider = ({ children }) => {
  const [query, setQuery] = useState('Subdivisions');

  return (
    <SearchContext.Provider value={{ query, setQuery }}>
      {children}
    </SearchContext.Provider>
  );
};

export { useSearch, SearchContextProvider as default };
