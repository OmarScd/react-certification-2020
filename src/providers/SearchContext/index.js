export { default, useSearch } from './SearchContext.provider';
