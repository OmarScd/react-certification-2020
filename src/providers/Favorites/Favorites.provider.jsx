/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect, useContext, useReducer } from 'react';

import { USERNAME_STORAGE_KEY } from '../../utils/constants';
import { storage } from '../../utils/storage';
import { useAuth } from '../Auth';
import favoritesReducer from '../../reducers/favorites';

const FavoritesContext = React.createContext([]);

const useFavorites = () => {
  const context = useContext(FavoritesContext);
  if (!context) {
    throw new Error(`Can't use "useFavorites" without a FavoritesProvider!`);
  }
  return context;
};

const FavoritesProvider = ({ children }) => {
  const [favorites, favoritesDispatch] = useReducer(favoritesReducer, []);
  const { username, authenticated } = useAuth();
  const userFavoritesStorageKey = `${USERNAME_STORAGE_KEY}-${username}`;

  useEffect(() => {
    if (authenticated) {
      const favoritesData = storage.get(userFavoritesStorageKey);
      if (favoritesData) {
        favoritesDispatch({ type: 'POPULATE_FAVORITES', favorites: favoritesData });
      }
    } else {
      favoritesDispatch({ type: 'CLEAR_FAVORITES' });
    }
  }, [authenticated]);

  useEffect(() => {
    if (authenticated) {
      storage.set(userFavoritesStorageKey, favorites);
    }
  }, [favorites]);

  return (
    <FavoritesContext.Provider value={{ favorites, favoritesDispatch }}>
      {children}
    </FavoritesContext.Provider>
  );
};

export { useFavorites, FavoritesProvider as default };
