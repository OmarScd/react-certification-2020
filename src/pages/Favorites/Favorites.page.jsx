import React from 'react';
import styled from 'styled-components';

import Layout from '../../components/Layout';
import VideoCard from '../../components/VideoCard';
import { useFavorites } from '../../providers/Favorites';

const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin: 0 auto;
  max-width: 1500px;
  padding: 0 3rem;
  justify-content: center;
`;

const FavoritesPage = () => {
  const { favorites } = useFavorites();

  return (
    <Layout>
      <Container>
        {favorites.length > 0 ? (
          favorites.map((item) => (
            <VideoCard
              key={item.id}
              id={item.id}
              thumbnail={item.thumbnail}
              title={item.title}
              description={item.description}
            />
          ))
        ) : (
          <p>You don&apos;t have any favorites yet</p>
        )}
      </Container>
    </Layout>
  );
};

export default FavoritesPage;
