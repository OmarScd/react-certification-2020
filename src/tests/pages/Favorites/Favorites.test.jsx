import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Favorites from '../../../pages/Favorites';
import ThemeProvider from '../../../providers/Theme';
import '../../../providers/SearchContext';
import '../../../providers/Auth';
import '../../../providers/Favorites';

jest.mock('../../../providers/SearchContext', () => ({
  default: jest.fn(),
  useSearch: () => ({ query: 'test', setQuery: jest.fn() }),
}));

jest.mock('../../../providers/Auth', () => ({
  default: jest.fn(),
  useAuth: () => ({ authenticated: true }),
}));

jest.mock('../../../providers/Favorites', () => ({
  default: jest.fn(),
  useFavorites: () => {
    return {
      favorites: [
        {
          id: 'YNZmZ4ARr38',
          thumbnail: 'https://i.ytimg.com/vi/YNZmZ4ARr38/mqdefault.jpg',
          title:
            'Shrek Forever After (2010) - Daddy Ever After Scene (2/10) | Movieclips',
          description:
            'Shrek Forever After - Daddy Ever After: Shrek (Mike Myers) begins to feel the monotony of family life with Donkey (Eddie Murphy) and Fiona (Cameron Diaz).',
        },
        {
          id: '0L1sL54G45Q',
          thumbnail: 'https://i.ytimg.com/vi/0L1sL54G45Q/mqdefault.jpg',
          title:
            'Shrek the Third (2007) - Damsels of Destruction Scene (8/10) | Movieclips',
          description:
            'Shrek the Third - Damsels of Destruction: The princesses break into the castle. BUY THE MOVIE: ...',
        },
        {
          id: 'mFl8nzZuExE',
          thumbnail: 'https://i.ytimg.com/vi/mFl8nzZuExE/mqdefault.jpg',
          title: 'Shrek (2001) - Do You Know the Muffin Man? Scene (2/10) | Movieclips',
          description:
            'Shrek - Do You Know the Muffin Man?: Lord Farquad (John Lithgow) interrogates the Gingerbread Man (Conrad Vernon) as to the whereabouts of the other ...',
        },
      ],
    };
  },
}));

describe('Favorites page', () => {
  afterAll(cleanup);

  it('renders correctly', () => {
    const { container } = render(
      <ThemeProvider>
        <BrowserRouter>
          <Favorites />
        </BrowserRouter>
      </ThemeProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
