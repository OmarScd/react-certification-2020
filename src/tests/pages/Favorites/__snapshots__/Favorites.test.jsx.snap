// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites page renders correctly 1`] = `
<div
  class="container"
>
  <header
    class="sc-hKFyIo jcdTAl"
  >
    <div
      class="sc-eCApGN hCdAiT"
    >
      <div
        class="sc-jSFkmK qmXvl"
      >
        <a
          class="sc-gKAblj iUhdyg"
          href="/"
        >
          Home
        </a>
      </div>
      <div
        class="sc-jSFkmK qmXvl"
      >
        <a
          class="sc-gKAblj iUhdyg"
          href="/favorites"
        >
          Favorites
        </a>
      </div>
      <div
        class="sc-jSFkmK qmXvl"
      >
        <form
          class="sc-bdnylx hYwFxl"
        >
          <img
            alt="magnifying lens"
            class="sc-dlnjPT dGgzTM"
            src="search-24px.svg"
          />
          <input
            class="sc-gtssRu fkigkU"
            placeholder="Search..."
            type="text"
            value="test"
          />
        </form>
      </div>
    </div>
    <div
      class="sc-eCApGN hCdAiT"
    >
      <div
        class="sc-jSFkmK qmXvl"
      >
        <button
          class="sc-iCoHVE kHdmrf"
        >
          Log out
        </button>
      </div>
      <div
        class="sc-jSFkmK qmXvl"
      >
        <img
          alt="person"
          class="sc-fujyUd hfIOzZ"
          src="person-24px.svg"
        />
      </div>
    </div>
  </header>
  <div
    class="content"
  >
    <div
      class="sc-bqGHjH fyLfWx"
    >
      <div
        class="sc-jrsJCI hISZqV"
      >
        <a
          class="sc-kEqYlL dOZYvC"
          href="/video/YNZmZ4ARr38"
        >
          <img
            alt="Thumbnail, no description available"
            class="sc-iqAbSa dnQfbv"
            src="https://i.ytimg.com/vi/YNZmZ4ARr38/mqdefault.jpg"
          />
        </a>
        <button
          class="sc-pNWxx diBcka"
        >
          Remove from
           favorites
        </button>
        <a
          class="sc-kEqYlL dOZYvC"
          href="/video/YNZmZ4ARr38"
        >
          <h2
            class="sc-crzoUp hrKUm"
          >
            Shrek Forever After (2010) - Daddy Ever After Scene (2/10) | Movieclips
          </h2>
          <p
            class="sc-dIsAE xqXss"
          >
            Shrek Forever After - Daddy Ever After: Shrek (Mike Myers) begins to feel the monotony of family life with Donkey (Eddie Murphy) and Fiona (Cameron Diaz).
          </p>
        </a>
      </div>
      <div
        class="sc-jrsJCI hISZqV"
      >
        <a
          class="sc-kEqYlL dOZYvC"
          href="/video/0L1sL54G45Q"
        >
          <img
            alt="Thumbnail, no description available"
            class="sc-iqAbSa dnQfbv"
            src="https://i.ytimg.com/vi/0L1sL54G45Q/mqdefault.jpg"
          />
        </a>
        <button
          class="sc-pNWxx diBcka"
        >
          Remove from
           favorites
        </button>
        <a
          class="sc-kEqYlL dOZYvC"
          href="/video/0L1sL54G45Q"
        >
          <h2
            class="sc-crzoUp hrKUm"
          >
            Shrek the Third (2007) - Damsels of Destruction Scene (8/10) | Movieclips
          </h2>
          <p
            class="sc-dIsAE xqXss"
          >
            Shrek the Third - Damsels of Destruction: The princesses break into the castle. BUY THE MOVIE: ...
          </p>
        </a>
      </div>
      <div
        class="sc-jrsJCI hISZqV"
      >
        <a
          class="sc-kEqYlL dOZYvC"
          href="/video/mFl8nzZuExE"
        >
          <img
            alt="Thumbnail, no description available"
            class="sc-iqAbSa dnQfbv"
            src="https://i.ytimg.com/vi/mFl8nzZuExE/mqdefault.jpg"
          />
        </a>
        <button
          class="sc-pNWxx diBcka"
        >
          Remove from
           favorites
        </button>
        <a
          class="sc-kEqYlL dOZYvC"
          href="/video/mFl8nzZuExE"
        >
          <h2
            class="sc-crzoUp hrKUm"
          >
            Shrek (2001) - Do You Know the Muffin Man? Scene (2/10) | Movieclips
          </h2>
          <p
            class="sc-dIsAE xqXss"
          >
            Shrek - Do You Know the Muffin Man?: Lord Farquad (John Lithgow) interrogates the Gingerbread Man (Conrad Vernon) as to the whereabouts of the other ...
          </p>
        </a>
      </div>
    </div>
  </div>
</div>
`;
